Sources/SwiftFusion/Inference/BearingRangeFactor.swift

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+
+/// Protocol for indicating a struct is range-capable
+public protocol RangeFunction: Differentiable {
+  associatedtype Base: Differentiable
+  associatedtype Target: Differentiable
+  @differentiable
+  static func range(_ from: Base, _ to: Target) -> Double
+}
+
+public struct PointToPointRange2D: RangeFunction {
+  @differentiable
+  public static func range(_ from: Vector2, _ to: Vector2) -> Double {
+    (to - from).norm
+  }
+}
+
+/// Protocol for indicating a struct is bearing-capable
+public protocol BearingFunction: Differentiable {
+  associatedtype Base: Differentiable
+  associatedtype Target: Differentiable
+  associatedtype Bearing: Differentiable
+  @differentiable
+  static func bearing(_ from: Base, _ to: Target) -> Bearing
+}
+
+public struct PoseToPointBearingRange2D: RangeFunction & BearingFunction {
```

Comment:
Split into Pose2ToVector2Bearing and Pose2ToVector2Range and move to Pose2

Reply:
See previous comment

Reply:
Still. Move to Pose2 and make it an extension of Pose2?

Reply:
It can't be an extension of Pose2 because this is to mimic the C++ `trait` behavior, as a protocol can only have one implementation in Swift.

Reply:
Still rename to Pose2ToVector2BearingRange and move to Pose2?

Reply:
Sure, done :)